test/regress_dns.c

```diff
@@ -2114,6 +2115,136 @@ test_getaddrinfo_async(void *arg)
 	test_ai_eq(a_out[12].ai, "18.52.86.120:8000", SOCK_STREAM, IPPROTO_TCP);
 	tt_str_op(a_out[12].ai->ai_canonname, ==, HOST_NAME_MAX_NAME);
 
+	/* 3. Let's make sure the results are all cached */
+
+	n_gai_results_pending = 13;
+
+	/* 0: both.example.com should have been replaced (evicted) in cache with no canonname */
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = PF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_flags = EVUTIL_AI_CANONNAME;
+	r = evdns_getaddrinfo(dns_base, "both.example.com", "8000",
+	    &hints, gai_cb, &b_out[0]);
+	tt_assert(r);
+
+	/* 1: v4only.example.com should have been cached, but CNAME was not replied with and the client wants CNAME. */
+	// XXX ideally, cache should hit with no CNAME if AI_CANONNAME on the previous call didn't obtain it with this flag.
+	hints.ai_flags = AI_CANONNAME;
+	r = evdns_getaddrinfo(dns_base, "v4only.example.com", "8001",
+	    &hints, gai_cb, &b_out[1]);
+	if (!r) {
```

Comment:
In which case the request may not be served from cache?

Reply:
The XXX comment just above explains it. We can add it in if that seems reasonable to you.
If the client requests AI_CANONNAME but server hasn't replied CNAME, it triggers a cache miss.

Reply:
But we have deterministic behavior of our server in tests, ti does not replies with CNAME for `v4only.examp.ecom`

Reply:
Exactly. Because v4only does not reply with CNAME, it is cached as having none. When we ask with AI_CANONNAME, we have two options:

1. Send the query again, maybe this time there will be a CNAME
2. Assume that it's not going to be replied and return ai with NULL canonname.

The original PR doesn't implement this so I stuck for the former, but if you prefer 2 I can change it.

Reply:
Hmm ok let's do this [10b3607](https://github.com/libevent/libevent/pull/1717/commits/10b3607c3fd3fbdc46c50fa722705c88e13e55bc)